### PR TITLE
[kubelet] Add pod cgroups support

### PR DESCRIPTION
**Pull request by dubstack**
This PR is linked to the upstream issue #27204 for introducing pod level cgroups into Kubernetes.

@derekwaynecarr  @vishh @Random-Liu PTAL.

I would like some suggestion/discussion on some comments that I would add in the PR.
I would also be adding e2e tests to this PR.
Please note that only the second commit is unique to this PR.

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/29087)

### CODE REVIEW

File: pkg/kubelet/cm/helpers_linux.go [hunk 1]
```diff
@@ -20,39 +20,84 @@ import (
...
+		MountPoints: mountPoints,
 	}, nil
 }
+
+// GetPodResourceRequest returns the pod requests for the supported resources.
+// Pod request is the summation of resource requests of all containers in the pod.
+func GetPodResourceRequests(pod *api.Pod) api.ResourceList {
+	requests := api.ResourceList{}
+	zeroQuantity := resource.MustParse("0")
+	for _, resource := range []api.ResourceName{api.ResourceCPU, api.ResourceMemory} {
```
Comment (derekwaynecarr) [inline pkg/kubelet/cm/helpers_linux.go]: we have utilities that support adding resource lists together that would simplify this code.

https://github.com/kubernetes/kubernetes/blob/master/pkg/quota/resources.go#L65

File: pkg/kubelet/cm/helpers_linux.go [hunk 2]
```diff
@@ -20,39 +20,84 @@ import (
...
+				if _, exists := limits[resource]; !exists {
+					limits[resource] = *delta
+				} else {
+					delta.Add(limits[resource])
+					limits[resource] = *delta
+				}
+			} else {
+				// if limit not specified for a particular resource in a container
+				// we default the pod resource limit to the resource capacity of the node
+				if cap, exists := capacity[resource]; exists {
```
Comment (derekwaynecarr) [inline pkg/kubelet/cm/helpers_linux.go]: does this mean we will show limits whose reported value is greater than node capacity if a pod has 2 containers and ctr1 has limits, and ctr2 does not?

Comment (derekwaynecarr) [inline pkg/kubelet/cm/helpers_linux.go]: ignore, missed the break

File: pkg/kubelet/cm/helpers_linux.go (hunk 1 above)
Comment (dubstack) [inline pkg/kubelet/cm/helpers_linux.go]: Thanks!

**Comment by dubstack** [comment]
@Random-Liu @derekwaynecarr This PR is ready for review.

File: pkg/kubelet/cm/helpers_linux.go [hunk 3]
```diff
@@ -44,3 +47,46 @@ func GetCgroupSubsystems() (*CgroupSubsystems, error) {
 		MountPoints: mountPoints,
 	}, nil
 }
+
+// GetPodResourceRequest returns the pod requests for the supported resources.
+// Pod request is the summation of resource requests of all containers in the pod.
+func GetPodResourceRequests(pod *api.Pod) api.ResourceList {
```
Comment (vishh) [inline pkg/kubelet/cm/helpers_linux.go]: nit: Why not re-use https://github.com/kubernetes/kubernetes/blob/master/pkg/api/resource_helpers.go#L165 or better refactor that and use it here?

File: pkg/kubelet/cm/helpers_linux.go [hunk 4]
```diff
@@ -44,3 +47,46 @@ func GetCgroupSubsystems() (*CgroupSubsystems, error) {
...
+	for _, container := range pod.Spec.Containers {
+		requests = quota.Add(requests, container.Resources.Requests)
+	}
+	return requests
+}
+
+// GetPodResourceLimits returns the pod limits for the supported resources
+// Pod limit is the summation of resource limits of all containers
+// in the pod. If limit for a particular resource is not specified for
+// even a single container then we return the node resource capacity
```
Comment (vishh) [inline pkg/kubelet/cm/helpers_linux.go]: Note that we should be returning Allocatable instead..

File: pkg/kubelet/cm/helpers_linux.go [hunk 5]
```diff
@@ -44,3 +47,46 @@ func GetCgroupSubsystems() (*CgroupSubsystems, error) {
 		MountPoints: mountPoints,
 	}, nil
 }
+
+// GetPodResourceRequest returns the pod requests for the supported resources.
+// Pod request is the summation of resource requests of all containers in the pod.
+func GetPodResourceRequests(pod *api.Pod) api.ResourceList {
+	requests := api.ResourceList{}
+	for _, container := range pod.Spec.Containers {
+		requests = quota.Add(requests, container.Resources.Requests)
```
Comment (vishh) [inline pkg/kubelet/cm/helpers_linux.go]: @derekwaynecarr 
It is odd that helper methods are in `quota` package. Should we move them to `pkg/api`, now that it is being used outside of quota pkg?

File: pkg/kubelet/cm/helpers_linux.go [hunk 6]
```diff
@@ -44,3 +47,46 @@ func GetCgroupSubsystems() (*CgroupSubsystems, error) {
...
+// in the pod. If limit for a particular resource is not specified for
+// even a single container then we return the node resource capacity
+// as the pod limit for the particular resource.
+func GetPodResourceLimits(pod *api.Pod, nodeInfo *api.Node) api.ResourceList {
+	capacity := nodeInfo.Status.Capacity
+	limits := api.ResourceList{}
+	zeroQuantity := resource.MustParse("0")
+	for _, resource := range []api.ResourceName{api.ResourceCPU, api.ResourceMemory} {
+		for _, container := range pod.Spec.Containers {
+			quantity := container.Resources.Limits[resource]
```
Comment (vishh) [inline pkg/kubelet/cm/helpers_linux.go]: You could also check if the quantity exists first before checking if its zero. I'm not sure if our defaulting logic adds a zero quantity whenever limits are not specified. In any case, it is better to stick to golang semantics for maps. Check that a value exists before using it.

File: pkg/kubelet/cm/helpers_linux.go [hunk 7]
```diff
@@ -44,3 +47,46 @@ func GetCgroupSubsystems() (*CgroupSubsystems, error) {
...
+			if quantity.Cmp(zeroQuantity) == 1 {
```
Comment (vishh) [inline pkg/kubelet/cm/helpers_linux.go]: Use `quantity.IsZero()` instead.

**Comment by vishh** [comment]
@dubstack Address my comments and hopefully this is ready to be merged.

File: pkg/kubelet/cm/helpers_linux.go (hunk 7 above)
Comment (dubstack) [inline pkg/kubelet/cm/helpers_linux.go]: Ack!

File: pkg/kubelet/cm/helpers_linux.go (hunk 4 above)
Comment (dubstack) [inline pkg/kubelet/cm/helpers_linux.go]: Ack!

File: pkg/kubelet/cm/helpers_linux.go (hunk 6 above)
Comment (dubstack) [inline pkg/kubelet/cm/helpers_linux.go]: Ack!

**Comment by dubstack** [comment]
@vishh Comments addressed. PTAL.

**Comment by dubstack** [comment]
Rebased on #28017.

File: pkg/kubelet/cm/helpers_linux.go (hunk 5 above)
Comment (derekwaynecarr) [inline pkg/kubelet/cm/helpers_linux.go]: I was discussing this with @smarterclayton and @deads2k recently that they could move elsewhere.  I am fine to move those in a follow-on (basically, everything but `CalculateUsage`) in that resources.go.  I can look to do that.

**Comment by dubstack** [comment]
@vishh @derekwaynecarr I believe you prefer having the nodeallocatableProvider interface in this PR itself? I will add the same. 

@Random-Liu Please review for other comments.

**Comment by dubstack** [comment]
I signed it.

**Comment by vishh** [comment]
@dubstack Thanks for continuing to work on this. Please ping me once the PR is ready for review.

**Comment by dubstack** [comment]
This PR would have to rebased on top of #30475, apart from that this should be good.

File: pkg/kubelet/cm/pod_container_manager_linux.go [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+	}
+	podContainer := podCgroupNamePrefix + string(pod.UID)
+	// Get the absolute path of the cgroup
+	return path.Join(parentContainer, podContainer)
+}
+
+// Destroy destroys the pod container cgroup paths
+func (m *podContainerManagerImpl) Destroy(podCgroup string) error {
+	// This will house the logic for destroying the pod cgroups.
+	// Will be handled in the next PR.
```
Comment (vishh) [inline pkg/kubelet/cm/pod_container_manager_linux.go]: @dubstack Why not delete the pod cgroups in this PR itself? Given that we decided to add a simple delete?
Is there a separate PR in the queue that handles deletion?

**Comment by vishh** [comment]
@dubstack Just one question on deletion. Otherwise this PR lgtm

**Comment by dubstack** [comment]
Commit from this PR, to be merged with #29049. Closing this one.

**Pull request closed without merging**